libnetwork/internal/l2disco/unsol_arp_linux.go

```diff
@@ -0,0 +1,79 @@
+package l2disco
+
+import (
+	"encoding/binary"
+	"fmt"
+	"net"
+	"slices"
+
+	"golang.org/x/sys/unix"
+)
+
+var (
+	arpTemplate = []byte{
+		0x00, 0x01, // Hardware type
+		0x08, 0x00, // Protocol
+		0x06,       // Hardware address length
+		0x04,       // IPv4 address length
+		0x00, 0x01, // ARP request
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Sender MAC
+		0x00, 0x00, 0x00, 0x00, // Sender IP
+		0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Target MAC (always zeros)
+		0x00, 0x00, 0x00, 0x00, // Target IP
+	}
+	bcastMAC = []byte{0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
+)
+
+type UnsolARP struct {
+	pkt []byte
+	sd  int
+	sa  *unix.SockaddrLinklayer
+}
+
+// NewUnsolARP returns a pointer to an object that can send unsolicited ARPs on
+// the interface with ifIndex, for ip and mac.
+func NewUnsolARP(ip net.IP, mac net.HardwareAddr, ifIndex int) (*UnsolARP, error) {
+	sd, err := unix.Socket(unix.AF_PACKET, unix.SOCK_DGRAM|unix.SOCK_CLOEXEC, int(htons(unix.ETH_P_ARP)))
+	if err != nil {
+		return nil, fmt.Errorf("create socket: %w", err)
+	}
+
+	pkt := slices.Clone(arpTemplate)
+	copy(pkt[8:14], mac)
+	copy(pkt[14:18], ip)
+	copy(pkt[24:28], ip)
+
+	sa := &unix.SockaddrLinklayer{
+		Protocol: htons(unix.ETH_P_ARP),
+		Ifindex:  ifIndex,
+		Hatype:   unix.ARPHRD_ETHER,
+		Halen:    uint8(len(bcastMAC)),
+	}
+	copy(sa.Addr[:], bcastMAC)
+
+	return &UnsolARP{
+		pkt: pkt,
+		sd:  sd,
+		sa:  sa,
+	}, nil
+}
+
+// Send sends an unsolicited ARP message.
+func (ua *UnsolARP) Send() error {
```

Comment:
Because it makes the socket's lifecycle very simple ... create, send messages, close.

With a socket per namespace, either the socket has to be created and kept around until the container's deleted, or it needs to be closed when ARPs have been sent for all interfaces (then another socket needs to be opened if the container is connected to another network). It also means the `Namespace` has some extra state to track.

I'm not convinced it's worthwhile - so, I've put it into a separate commit. Can squash it, remove, or fiddle around some more.